File: slideWindow/FindAnagrams438.java
public class FindAnagrams438{
    public static void main(String[] args) {


    }

    /*
            输入:
                s: "cbaebabacd" p: "abc"
            输出:
                [0, 6]
            解释:
                起始索引等于 0 的子串是 "cba", 它是 "abc" 的字母异位词。
                起始索引等于 6 的子串是 "bac", 它是 "abc" 的字母异位词。
    */
    public List<Integer> findAnagrams1(String s, String p) {
        int[] target=new int[256];
        int[] window=new int[256];
        int l=0,r=0;
        int plen=p.length();
        int slen=s.length();
        int count=0,match=0;
        List<Integer> res=new ArraysList<>();
        for (int i=0;i<plen;i++) {
            target[p.charAt(i)]++;
        }
        for (int a:target) {
            if(a!=0){
                count++;
            }
        }

        while(r<slen){
            char right=s.charAt(r);
            if (target[right]!=0) {
                window[right]++;
                if (window[right]==target[right]) {
                    match++;
                }
            }
            r++;   
            while(l<r && count==match){
                char left=s.charAt(l);
                if (r-l==res) {
                    res.add(l);
                }
                l++;
                //不满足了
                if (target[left]!=0) {
                    window[left]--;
                    if (window[left]<target[left]) {
                        match--;
                    }
                }
            }
        }
        return res;
    }

    //wrong
    public List<Integer> findAnagrams2(String s, String p) {
        int[] target=new int[256];
        int[] window=new int[256];
        int l=0,r=0;
        int plen=p.length();
        int slen=s.length();
        int count=0,match=0;
        List<Integer> res=new ArraysList<>();
        for (int i=0;i<plen;i++) {
            target[p.charAt(i)]++;
        }
        for (int a:target) {
            if(a!=0){
                count++;
            }
        }

        while(r<slen){
            char right=s.charAt(r);
            if (target[right]!=0) {
                window[right]++;
                if (window[right]==target[right]) {
                    match++;
                }
                r++;
            }else{
                if (match==count) {
                    res.add(l);
                }
                l=r+1;
                r=l;
                match=0;
            }
        }
        return res;
    }
}

File: slideWindow/FindAnagrams438.go
func findAnagrams(s string, p string) []int {
    var target = make([]int, 128)
    var window = make([]int, 128)
    var match = 0
    for _, sp := range p {
        if target[sp] == 0 {
            match++
        }
        target[sp]++
    }
    var left = 0
    var count = 0
    var res []int
    for right := 0; right < len(s); right++ {
        if target[s[right]] > 0 {
            window[s[right]]++
            if window[s[right]] == target[s[right]] {
                count++
            }
        }
        for count == match && left <= right {
            if right-left+1 == len(p) {
                res = append(res, left)
            }
            if window[s[left]] > 0 {
                window[s[left]]--
                if window[s[left]] < target[s[left]] {
                    count--
                }
            }
            left++
        }
    }
    return res
}

func findAnagrams(s string, p string) []int {
    var left = 0
    //-'a'看起来太丑了，直接128
    var target [128]int //注意用数组，可以直接比较
    var window [128]int
    for _, sp := range p {
        target[sp]++
    }
    var res []int
    for right := 0; right < len(s); right++ {
        if target[s[right]] > 0 {
            window[s[right]]++
        }
        for right-left+1 > len(p) {
            if window[s[left]] > 0 {
                window[s[left]]--
            }
            left++
        }
        if window == target {
            res = append(res, left)
        }
    }
    return res
}
